server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java

```diff
@@ -203,8 +203,9 @@ void writeMetadataBlob(String blobName, boolean failIfAlreadyExists, boolean ato
      * @param expected the expected value
      * @param updated the new value
      * @return the value read from the register (before it was updated)
+     * @throws ConcurrentRegisterOperationException if concurrent activity prevents us from even reading the value of the register
      */
-    long compareAndExchangeRegister(String key, long expected, long updated) throws IOException;
+    long compareAndExchangeRegister(String key, long expected, long updated) throws IOException, ConcurrentRegisterOperationException;
```

Comment:
nit: does it need to be a checked exception?

Reply:
No, but I expect that callers will need to handle this exception very close to the call site (much like handling the case where the operation failed but managed to read the value) and forgetting to do this would be pretty bad, so I went this way. Perhaps it would be nicer to return an `OptionalLong`, we don't really need any details?

Reply:
> Perhaps it would be nicer to return an OptionalLong, we don't really need any details?

see https://github.com/elastic/elasticsearch/commit/ab1385b8c5237dcec3e6aff0ecb36682fa1e91d4 for a draft of that